fastestimator/trace/metric/bleu_score.py

```diff
@@ -0,0 +1,174 @@
+        self.no_of_correct_predicted = Counter()
+        self.no_of_total_predicted = Counter()
+        self.total_hypotheses_length = 0
+        self.total_references_length = 0
+
+    def get_bleu_score(self):
+
+        # Calculate corpus-level brevity penalty.
+        bp = brevity_penalty(self.total_references_length, self.total_hypotheses_length)
+
+        # Collects the various precision values for the different ngram orders.
+        p_n = [
+            Fraction(self.no_of_correct_predicted[i], self.no_of_total_predicted[i], _normalize=False)
+            for i in range(1, self.n_gram + 1)
+        ]
+
+        # Returns 0 if there's no matching 1-gram
+        if self.no_of_correct_predicted[1] == 0:
+            return 0
+
+        # Smoothen the modified precision.
+        p_n = self.smoothing_function(p_n, [], [], hyp_len=self.total_hypotheses_length)
+
+        geometric_average_precision = math.exp(
+            math.fsum((w_i * math.log(p_i) for w_i, p_i in zip(self.weights, p_n) if p_i > 0)))
+        bleu_score = bp * geometric_average_precision
+
+        return bleu_score
+
+    def batch_precision_parameters(self, list_of_references, hypotheses):
+
+        assert len(list_of_references) == len(hypotheses), (
+            "The number of hypotheses and their reference(s) should be the same ")
+
+        # Iterate through each hypothesis and their corresponding references.
+        for references, hypothesis in zip(list_of_references, hypotheses):
+
+            # For each order of ngram, calculate the correct predicted words and
+            # total predicted words for the corpus-level modified precision.
+            references = get_formated_reference(references)
+            hypothesis = get_formated_hypotheses(hypothesis)
+            for i in range(1, self.n_gram + 1):
+                p_i = modified_precision(references, hypothesis, i)
+                self.no_of_correct_predicted[i] += p_i.numerator
+                self.no_of_total_predicted[i] += p_i.denominator
+
+            # Calculate the hypothesis length and the closest reference length.
+            # Adds them to the corpus-level hypothesis and reference counts.
+            hyp_len = len(hypothesis)
+            self.total_hypotheses_length += hyp_len
+            ref_lens = (len(reference) for reference in references)
+            self.total_references_length += min(ref_lens, key=lambda ref_len: (abs(ref_len - hyp_len), ref_len))
+
+    def on_batch_end(self, data: Data) -> None:
+        y_pred, y_true = to_number(data['pred']), to_number(data['target_real'])
+        if y_true.shape[-1] > 1 and y_true.ndim > 2:
+            y_true = np.argmax(y_true, axis=-1)
+        if y_pred.shape[-1] > 1 and y_pred.ndim > 2:
+            y_pred = np.argmax(y_pred, axis=-1)
+        self.batch_precision_parameters(y_true, y_pred)
+
+    def on_epoch_end(self, data: Data) -> None:
+        data[self.outputs[0]] = round(self.get_bleu_score(), 5)
```

Comment:
can we have a unit test demonstrating that the process of using this trace over several steps is equivalent to directly invoking the nltk corpus function?

Reply:
Already the code using corpus function and the results are the same. But will add a unit test to test the same.

Reply:
cool, it's good to have validation tests on the metrics since I believe this is a requirement during FDA submissions